src/pool/srv_util.c

```diff
@@ -202,91 +202,624 @@
+{
+	/* y_ranks may point to a static initializer; do not change it. */
+	d_rank_list_sort(x);
+	if (x->rl_nr != y_ranks_len)
+		return false;
+	if (memcmp(x->rl_ranks, y_ranks, x->rl_nr) != 0)
+		return false;
+	return true;
+}
+
+static bool
+testu_rank_sets_belong(d_rank_list_t *x, d_rank_t *y_ranks, int y_ranks_len)
+{
+	int i;
+
+	for (i = 0; i < x->rl_nr; i++) {
+		d_rank_t	rank = x->rl_ranks[i];
+		int		j;
+
+		for (j = 0; j < y_ranks_len; j++)
+			if (y_ranks[j] == rank)
+				break;
+		if (j == y_ranks_len)
+			return false;
 	}
+	return true;
+}
 
-	if (nreplaced < nfailed) {
-		D_WARN("Not enough ranks available; Failed %d, Replacements %d",
-			nfailed, nreplaced);
-		alt->rl_nr = nreplaced;
-		replicas->rl_nr -= (nfailed - nreplaced);
+static struct pool_map *
+testu_create_pool_map(d_rank_t *ranks, int n_ranks, d_rank_t *down_ranks, int n_down_ranks)
+{
+	d_rank_list_t		ranks_list = {
+		.rl_ranks	= ranks,
+		.rl_nr		= n_ranks
+	};
+	struct pool_buf	       *map_buf;
+	struct pool_map	       *map;
+	uint32_t	       *domains;
+	int			n_domains = 3 + n_ranks;
+	int			i;
+	int			rc;
+
+	/* Not using domains properly at the moment. See FD_TREE_TUNPLE_LEN. */
+	D_ALLOC_ARRAY(domains, n_domains);
+	D_ASSERT(domains != NULL);
+	domains[0] = 1;
+	domains[1] = 0;
+	domains[2] = n_ranks;
+	for (i = 0; i < n_ranks; i++)
+		domains[3 + i] = i;
+
+	rc = gen_pool_buf(NULL /* map */, &map_buf, 1 /* map_version */, n_domains,
+			  n_ranks, n_ranks * 1 /* ntargets */, domains, &ranks_list,
+			  1 /* dss_tgt_nr */);
+	D_ASSERT(rc == 0);
+
+	rc = pool_map_create(map_buf, 1, &map);
+	D_ASSERT(rc == 0);
+
+	for (i = 0; i < n_down_ranks; i++) {
+		struct pool_domain *d;
+
+		d = pool_map_find_node_by_rank(map, down_ranks[i]);
+		D_ASSERT(d != NULL);
+		d->do_comp.co_status = PO_COMP_ST_DOWN;
 	}
-	return 0;
+
+	pool_buf_free(map_buf);
+	D_FREE(domains);
+	return map;
+}
+
+static void
+testu_plan_svc_reconfs(int svc_rf, d_rank_t ranks[], int n_ranks, d_rank_t down_ranks[],
+		       int n_down_ranks, d_rank_t replicas_ranks[], int n_replicas_ranks,
+		       d_rank_t self, d_rank_list_t **to_add, d_rank_list_t **to_remove)
+{
+	struct pool_map	       *map;
+	d_rank_list_t		replicas_list;
+	int			rc;
+
+	map = testu_create_pool_map(ranks, n_ranks, down_ranks, n_down_ranks);
+
+	replicas_list.rl_ranks = replicas_ranks;
+	replicas_list.rl_nr = n_replicas_ranks;
+
+	rc = ds_pool_plan_svc_reconfs(svc_rf, map, &replicas_list, self, to_add, to_remove);
+	D_ASSERT(rc == 0);
+
+	pool_map_decref(map);
+}
+
+void
+ds_pool_test_plan_svc_reconfs(void)
+{
+	d_rank_t		self = 0;
+	d_rank_list_t	       *to_add;
+	d_rank_list_t	       *to_remove;
+
+#define call_testu_plan_svc_reconfs								\
+	testu_plan_svc_reconfs(svc_rf, ranks, ARRAY_SIZE(ranks), down_ranks,			\
+			       ARRAY_SIZE(down_ranks), replicas_ranks,				\
+			       ARRAY_SIZE(replicas_ranks), self, &to_add, &to_remove);
+
+#define call_d_rank_list_free									\
```

Comment:
(style)  Macros with multiple statements should be enclosed in a do - while loop
(style)  macros should not use a trailing semicolon